tools/codegen/src/generator/languages/rust_lazy_reader/generator.rs

```diff
@@ -0,0 +1,718 @@
+use super::{ident_new, LazyReaderGenerator};
+use crate::ast::{self, HasName, *};
+use case::CaseExt;
+use proc_macro2::{Literal, TokenStream};
+use quote::quote;
+use std::io;
+
+impl LazyReaderGenerator for ast::Union {
+    fn gen_rust<W: io::Write>(&self, output: &mut W) -> io::Result<()> {
+        let name = ident_new(self.name());
+
+        let q = quote! {
+            pub struct #name {
+                pub cursor: Cursor,
+            }
+
+            impl From<Cursor> for #name {
+                fn from(cursor: Cursor) -> Self {
+                    Self { cursor }
+                }
+            }
+
+            impl #name {
+                pub fn item_id(&self) -> Result<usize, Error> {
+                    let item = self.cursor.union_unpack()?;
+                    Ok(item.item_id)
+                }
+            }
+        };
+        writeln!(output, "{}", q)?;
+
+        for (item_index, item) in self.items().iter().enumerate() {
+            let item_type_name = item.typ().name();
+            let item_type_name = ident_new(&format!("as_{}", item_type_name.to_snake()));
+            let (transformed_name, tc) = get_rust_type_category(item.typ());
+            let convert_code = tc.gen_convert_code();
+            let q = quote! {
+                impl #name {
+                    pub fn #item_type_name(&self) -> Result<#transformed_name, Error> {
+                        let item = self.cursor.union_unpack()?;
+                        if item.item_id != #item_index {
+                            return Err(Error::Header(format!("invalid item type id in union: {}", item.item_id)));
+                        }
```

Comment:
Why we have `if item.item_id != #item_index` here?

Reply:
It is a common phenomenon that `item_index` and `item.index` are not equal. For example, in a union like the following:
```
union A {
 a: 2,
 b: 4,
}
```
the `item_index` and `item.index` are not equal.